src/rust/engine/src/rule_graph.rs

```diff
@@ -0,0 +1,583 @@
+    match self.entry_type {
+      EntryType::SubjectIsProduct => false,
+      EntryType::Literal => false,
+      EntryType::InnerEntry => true,
+      EntryType::Root => true,
+      EntryType::Unreachable => false,
+    }
+  }
+  fn can_be_dependency(&self) -> bool {
+    match self.entry_type {
+      EntryType::SubjectIsProduct => true,
+      EntryType::Literal => true,
+      EntryType::InnerEntry => true,
+      EntryType::Root => false,
+      EntryType::Unreachable => false,
+    }
+  }
+}
+
+type Entries = Vec<Entry>;
+type RootRuleDependencyEdges = HashMap<Entry, RuleEdges>; // Root -> InternalEntry
+type Rule = Task;
+type RuleDependencyEdges = HashMap<Entry, RuleEdges>; // InternalEntry -> InternalEntry
+type RuleDiagnostics = Vec<Diagnostic>;
+type UnfulfillableRuleMap = HashMap<Entry, RuleDiagnostics>;
+
+#[derive(Debug)]
+pub struct RootSubjectTypes<'a> {
+  pub subject_types: &'a Vec<TypeId>
+}
+
+#[derive(Eq, Hash, PartialEq, Clone, Debug)]
+pub struct Diagnostic {
+  subject_type: TypeId,
+  reason: String
+}
+
+/*
+ * Given the task index and the root subjects, it produces a rule graph that allows dependency nodes
+ * to be found statically rather than dynamically.
+ */
+pub struct GraphMaker<'a> {
+    tasks: Tasks,
+    root_subject_types: RootSubjectTypes<'a>
+}
+
+impl<'a> GraphMaker<'a> {
+  pub fn new(tasks: &Tasks, root_subject_types: RootSubjectTypes<'a>) -> GraphMaker<'a> {
+    let cloned_tasks = tasks.clone();
+    GraphMaker { tasks: cloned_tasks, root_subject_types: root_subject_types }
+  }
+
+  pub fn full_graph(&self) -> RuleGraph {
+    let mut full_root_rule_dependency_edges: RootRuleDependencyEdges = HashMap::new();
+    let mut full_dependency_edges: RuleDependencyEdges = HashMap::new();
+    let mut full_unfulfillable_rules: UnfulfillableRuleMap = HashMap::new();
+    // for root_subj_type, selectorfn in root_subj_selector_fn.items():...
+
+    //let product_types = self.tasks.all_product_types(&beginning_root.subject_type);
+    let product_types = self.tasks.all_product_types();
+    for beginning_root in self.gen_root_entries(&product_types) {
+      let constructed_graph = self._construct_graph(
+        beginning_root,
+        full_root_rule_dependency_edges,
+        full_dependency_edges,
+        full_unfulfillable_rules
+      );
+
+      // less than ideal, the copying
+      full_root_rule_dependency_edges = constructed_graph.root_dependencies.clone();
+      full_dependency_edges = constructed_graph.rule_dependency_edges.clone();
+      full_unfulfillable_rules = constructed_graph.unfulfillable_rules.clone();
+    }
+    let rules_in_graph: HashSet<Task> = full_dependency_edges.keys().map(|f| f.rule().clone()).collect();
+    let unfulfillable_discovered_during_construction: HashSet<Task> = full_unfulfillable_rules.keys().map(|f| f.rule().clone()).collect();
+    let declared_rules = self.tasks.all_rules();
+    let unreachable_rules: HashSet<&Task> = declared_rules.iter()
```

Comment:
Note that when the type annotation is use to guide inference for `collect`, it isn't necessary to actually fill in its parameters:
```
let unreachable_rules: HashSet<_> = declared_rules.iter()...
```